pkg/declcfg/write.go

```diff
@@ -0,0 +1,162 @@
+package declcfg
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/operator-framework/operator-registry/pkg/property"
+)
+
+const (
+	objectsDirName = "objects"
+	globalName     = "__global"
+)
+
+func WriteDir(cfg DeclarativeConfig, configDir string) error {
+	entries, err := ioutil.ReadDir(configDir)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if len(entries) > 0 {
+		return fmt.Errorf("config dir %q must be empty", configDir)
+	}
+
+	return writeToFS(cfg, &diskWriter{}, configDir)
+}
+
+type fsWriter interface {
+	MkdirAll(path string, mode os.FileMode) error
+	WriteFile(path string, data []byte, mode os.FileMode) error
+}
+
+var _ fsWriter = &diskWriter{}
+
+type diskWriter struct{}
+
+func (w diskWriter) MkdirAll(path string, mode os.FileMode) error {
+	return os.MkdirAll(path, mode)
+}
+
+func (w diskWriter) WriteFile(path string, data []byte, mode os.FileMode) error {
+	return ioutil.WriteFile(path, data, mode)
+}
+
+func writeToFS(cfg DeclarativeConfig, w fsWriter, rootDir string) error {
+	bundlesByPackage := map[string][]Bundle{}
+	for _, b := range cfg.Bundles {
+		bundlesByPackage[b.Package] = append(bundlesByPackage[b.Package], b)
+	}
+	othersByPackage := map[string][]Meta{}
+	for _, o := range cfg.Others {
+		pkgName := o.Package
+		if pkgName == "" {
+			pkgName = globalName
+		}
+		othersByPackage[pkgName] = append(othersByPackage[pkgName], o)
+	}
+
+	if err := w.MkdirAll(rootDir, 0755); err != nil {
+		return fmt.Errorf("mkdir %q: %v", rootDir, err)
+	}
+
+	for _, p := range cfg.Packages {
+		fcfg := DeclarativeConfig{
+			Packages: []Package{p},
+			Bundles:  bundlesByPackage[p.Name],
+			Others:   othersByPackage[p.Name],
+		}
+		pkgDir := filepath.Join(rootDir, p.Name)
+		if err := w.MkdirAll(pkgDir, 0755); err != nil {
+			return fmt.Errorf("mkdir %q: %v", pkgDir, err)
```

Comment:
Nit: `MkdirAll()` already returns an error with message prefix `mkdir <dir>:`.
```suggestion
			return err
```